builder/azure/arm/step_deploy_template.go

```diff
@@ -298,21 +286,48 @@ func (s *StepDeployTemplate) deleteDeploymentResources(ctx context.Context, subs
 	}
 
 	resources := map[string]string{}
-
+	var vmID *virtualmachines.VirtualMachineId
+	var nicID *commonids.NetworkInterfaceId
 	for _, deploymentOperation := range deploymentOperations.Items {
 		// Sometimes an empty operation is added to the list by Azure
 		if deploymentOperation.Properties.TargetResource == nil {
 			continue
 		}
-
 		resourceName := *deploymentOperation.Properties.TargetResource.ResourceName
 		resourceType := *deploymentOperation.Properties.TargetResource.ResourceType
 
-		s.say(fmt.Sprintf("Adding to deletion queue -> %s : '%s'", resourceType, resourceName))
-		resources[resourceType] = resourceName
-
+		// Create vm and NIC ID to use them to delete later
+		switch resourceType {
+		case "Microsoft.Compute/virtualMachines":
+			vmIDDeref := virtualmachines.NewVirtualMachineID(subscriptionId, resourceGroupName, resourceName)
+			vmID = &vmIDDeref
+		case "Microsoft.Network/networkInterfaces":
+			nicIDDeref := commonids.NewNetworkInterfaceID(subscriptionId, resourceGroupName, resourceName)
+			nicID = &nicIDDeref
+		default:
+			s.say(fmt.Sprintf("Adding to deletion queue -> %s : '%s'", resourceType, resourceName))
+			resources[resourceType] = resourceName
+		}
 	}
 
+	// Always delete VM first, then delete NIC, this way we avoid conflicts on deletion
+	if vmID != nil {
+		s.say(fmt.Sprintf("Attempting deletion -> %s : '%s'", "Microsoft.Compute/virtualMachines", vmID.VirtualMachineName))
+		if err := s.client.VirtualMachinesClient.DeleteThenPoll(pollingContext, *vmID, virtualmachines.DefaultDeleteOperationOptions()); err != nil {
+			s.reportIfError(err, vmID.VirtualMachineName)
```

Comment:
I wonder looking at this, why do we have a function called `reportIfError` if we conditionally call it only in case of error?

Reply:
@lbajolet-hashicorp in several places we don't call if it there is an error, I was just calling the same function to re-use the same warning message